csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -174,28 +272,78 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
   // examples. This is the reason behind `depended_by_non_aliases`.
   const std::unordered_set<Expr*>& depended_by_non_aliases =
       exprsDependedByNonAliases(analysis, fusion);
-  std::vector<Use> uses_to_segment;
-  uses_to_segment.reserve(fusion->outputs().size());
+  std::set<Use> uses_to_segment;
   for (auto* out : ir_utils::filterByType<TensorView>(fusion->outputs())) {
     Use use_to_segment =
         findUseToSegment(out, analysis, depended_by_non_aliases);
     if (use_to_segment.use_of != out) {
-      uses_to_segment.push_back(use_to_segment);
+      uses_to_segment.insert(use_to_segment);
     }
   }
 
-  // The remaining are optimizations to reduce the number of `segment_set`s
-  // inserted.
+  // Step 2: for inputs, the algorithm tries to walk down from each fusion
+  // input until reaching a non-meta or a fork. Stopping at the fork is to
+  // avoid feeding the same data via multiple inputs, e.g.,
   //
-  // Group `uses_to_segment` by `use_of` and remove duplicates.
-  std::sort(uses_to_segment.begin(), uses_to_segment.end());
-  uses_to_segment.erase(
-      std::unique(uses_to_segment.begin(), uses_to_segment.end()),
-      uses_to_segment.end());
+  //   in -> reshape_0 -> mul
+  //    |                  ^
+  //    +--> reshape_1 ----+
+  //
+  // If we separate `reshape_0` and `reshape_1` from `mul`, the pointwise
+  // kernel would take double the input.
```

Comment:
Just want to point out here that, `double the input` means double the number of TensorViews passed to the kernel. But in kernel execution, the redundant read might be saved from cache.
But I agree this is still an overall good thing to have in the reshape example.

~But this would stop us bookending multiple slicing on the same tensor.... arguably, in that case we aren't really duplicating memory buffer, if the slices aren't overlapping.~

~Not pushing to change this in the PR. Can we add a comment on that?~

NVM, I see you special case about `SliceOp` down below.